src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/DetectPreviewFeatureAnalyzer.cs

```diff
@@ -767,14 +878,160 @@ private static ISymbol SymbolFromArrayCreationOperation(IArrayCreationOperation
             return false;
         }
 
-        private static bool SymbolIsAnnotatedAsPreview(ISymbol symbol, ConcurrentDictionary<ISymbol, bool> requiresPreviewFeaturesSymbols, INamedTypeSymbol previewFeatureAttribute)
+#pragma warning disable CA1054,CA1055 // url should be of type URI
+        private static string? GetMessageAndURLFromAttributeConstructor(AttributeData attribute, out string? url)
+#pragma warning restore CA1054,CA1055
         {
-            if (!requiresPreviewFeaturesSymbols.TryGetValue(symbol, out bool existing))
+            string? message = null;
+            url = null;
+            ImmutableArray<TypedConstant> constructorArguments = attribute.ConstructorArguments;
+            if (constructorArguments.Length != 0)
             {
-                if (symbol.HasAttribute(previewFeatureAttribute))
+                if (constructorArguments.First().Value is string messageValue)
                 {
-                    requiresPreviewFeaturesSymbols.GetOrAdd(symbol, true);
-                    return true;
+                    message = messageValue;
+                }
+            }
+
+            ImmutableArray<System.Collections.Generic.KeyValuePair<string, TypedConstant>> namedArguments = attribute.NamedArguments;
+            if (namedArguments.Length != 0)
+            {
+                foreach (System.Collections.Generic.KeyValuePair<string, TypedConstant> namedArgument in namedArguments)
+                {
+                    if (namedArgument.Key == "Message")
+                    {
+                        message = (string)namedArgument.Value.Value;
+                    }
+
+                    if (namedArgument.Key == "Url")
+                    {
+                        url = (string)namedArgument.Value.Value;
+                    }
+                }
+            }
+
+            return message;
+        }
+
+        private static void ReportDiagnosticWithCustomMessageIfItExists(OperationAnalysisContext context,
+                                                                        IOperation operation,
+                                                                        ISymbol symbol,
+                                                                        ConcurrentDictionary<ISymbol, ValueTuple<bool, string?, string?>> requiresPreviewFeaturesSymbols,
```

Comment:
NIT: Using a named value tuple (`existing.url`) would look better than `existing.Item3`

```suggestion
                                                                        ConcurrentDictionary<ISymbol, (bool isPreview, string message, string url)> requiresPreviewFeaturesSymbols,
```

Reply:
yeah I agree